// @ngx-share/buttons/bundles/ngx-share-buttons.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@ngx-share/core'), require('rxjs/BehaviorSubject'), require('rxjs/operators/map'), require('@fortawesome/free-solid-svg-icons'), require('@angular/common'), require('@ngx-share/button')) :
	typeof define === 'function' && define.amd ? define('@ngx-share/buttons', ['exports', '@angular/core', '@ngx-share/core', 'rxjs/BehaviorSubject', 'rxjs/operators/map', '@fortawesome/free-solid-svg-icons', '@angular/common', '@ngx-share/button'], factory) :
	(factory((global['ngx-share'] = global['ngx-share'] || {}, global['ngx-share'].buttons = {}),global.ng.core,global.core$1,global.Rx,global.Rx.Observable.prototype,global.freeSolidSvgIcons,global.ng.common,global.button));
}(this, (function (exports,core,core$1,BehaviorSubject,map,freeSolidSvgIcons,common,button) { 'use strict';

var ShareButtonsComponent = (function () {
    function ShareButtonsComponent(share) {
        this.share = share;
        this.stateWorker$ = new BehaviorSubject.BehaviorSubject({
            includedButtons: [],
            excludedButtons: [],
            userButtons: [],
            selectedButtons: [],
            expanded: true,
            shownCount: Object.keys(this.share.config.prop).length
        });
        this.theme = this.share.theme;
        this.showIcon = true;
        this.showText = false;
        this.showCount = false;
        this.size = 0;
        this.count = new core.EventEmitter();
        this.opened = new core.EventEmitter();
        this.closed = new core.EventEmitter();
    }
    Object.defineProperty(ShareButtonsComponent.prototype, "includedButtons", {
        set: function (includedButtons) {
            this.updateState({ includedButtons: includedButtons });
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ShareButtonsComponent.prototype, "excludedButtons", {
        set: function (excludedButtons) {
            this.updateState({ excludedButtons: excludedButtons });
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ShareButtonsComponent.prototype, "shownButtons", {
        set: function (shownCount) {
            this.updateState({ shownCount: shownCount });
        },
        enumerable: true,
        configurable: true
    });
    ShareButtonsComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.state$ = this.stateWorker$.pipe(map.map(function (state) {
            var includedButtons = state.includedButtons.length ? state.includedButtons : state.userButtons;
            var userButtons = includedButtons.filter(function (btn) { return state.excludedButtons.indexOf(btn) < 0; });
            var selectedButtons = userButtons.slice(0, state.expanded ? userButtons.length : state.shownCount);
            return {
                userButtons: userButtons,
                selectedButtons: selectedButtons,
                expanded: state.expanded,
                shownCount: state.shownCount
            };
        }));
        this.configSub$ = this.share.config$.subscribe(function (config) {
            var includedButtons = config.options.include.length ? config.options.include : Object.keys(config.prop);
            var userButtons = includedButtons.filter(function (btn) { return config.options.exclude.indexOf(btn) < 0; });
            _this.updateState({
                userButtons: userButtons,
                expanded: false
            });
        });
    };
    ShareButtonsComponent.prototype.updateState = function (state) {
        this.stateWorker$.next(Object.assign({}, this.stateWorker$.getValue(), state));
    };
    ShareButtonsComponent.prototype.ngOnDestroy = function () {
        if (this.configSub$) {
            this.configSub$.unsubscribe();
        }
    };
    return ShareButtonsComponent;
}());
ShareButtonsComponent.decorators = [
    { type: core.Component, args: [{
                selector: 'share-buttons',
                template: "<div *ngIf=\"state$ | async; let state\" class=\"sb-group sb-{{theme}}\">\n  <share-button *ngFor=\"let button of state.selectedButtons\"\n                [button]=\"button\"\n                [theme]=\"theme\"\n                [url]=\"url\"\n                [title]=\"title\"\n                [description]=\"description\"\n                [image]=\"image\"\n                [tags]=\"tags\"\n                [autoSetMeta]=\"autoSetMeta\"\n                [showCount]=\"showCount\"\n                [showIcon]=\"showIcon\"\n                [showText]=\"showText\"\n                [size]=\"size\"\n                (opened)=\"opened.emit($event)\"\n                (closed)=\"closed.emit($event)\"\n                (count)=\"count.emit($event)\"></share-button>\n  <div class=\"sb-button sb-{{theme}}\">\n    <expand-button *ngIf=\"state.shownCount < state.userButtons.length\"\n                   [expanded]=\"state.expanded\"\n                   [size]=\"(1 + size/20) * 14\"\n                   (toggle)=\"updateState({expanded: $event})\">\n    </expand-button>\n  </div>\n</div>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                preserveWhitespaces: false
            },] },
];
ShareButtonsComponent.ctorParameters = function () { return [
    { type: core$1.ShareButtons, },
]; };
ShareButtonsComponent.propDecorators = {
    "theme": [{ type: core.Input },],
    "includedButtons": [{ type: core.Input, args: ['include',] },],
    "excludedButtons": [{ type: core.Input, args: ['exclude',] },],
    "shownButtons": [{ type: core.Input, args: ['show',] },],
    "url": [{ type: core.Input },],
    "title": [{ type: core.Input },],
    "description": [{ type: core.Input },],
    "image": [{ type: core.Input },],
    "tags": [{ type: core.Input },],
    "autoSetMeta": [{ type: core.Input },],
    "showIcon": [{ type: core.Input },],
    "showText": [{ type: core.Input },],
    "showCount": [{ type: core.Input },],
    "size": [{ type: core.Input },],
    "count": [{ type: core.Output },],
    "opened": [{ type: core.Output },],
    "closed": [{ type: core.Output },],
};
var ExpandButtonComponent = (function () {
    function ExpandButtonComponent(el) {
        this.icon = {
            true: freeSolidSvgIcons.faMinus,
            false: freeSolidSvgIcons.faEllipsisH
        };
        this.toggle = new core.EventEmitter();
        el.nativeElement.style.setProperty('--button-color', '#FF6651');
    }
    return ExpandButtonComponent;
}());
ExpandButtonComponent.decorators = [
    { type: core.Component, args: [{
                selector: 'expand-button',
                template: "\n    <button class=\"sb-wrapper sb-expand sb-show-icon\"\n            [style.fontSize.px]=\"size\"\n            (click)=\"toggle.emit(!expanded)\">\n      <div class=\"sb-inner\">\n        <div class=\"sb-content\">\n          <div class=\"sb-icon\">\n            <fa-icon [icon]=\"icon[expanded]\"></fa-icon>\n          </div>\n        </div>\n      </div>\n    </button>\n  ",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                preserveWhitespaces: false
            },] },
];
ExpandButtonComponent.ctorParameters = function () { return [
    { type: core.ElementRef, },
]; };
ExpandButtonComponent.propDecorators = {
    "expanded": [{ type: core.Input },],
    "size": [{ type: core.Input },],
    "toggle": [{ type: core.Output },],
};
function ShareButtonsFactory(config) {
    return new core$1.ShareButtons(config);
}
var ShareButtonsModule = (function () {
    function ShareButtonsModule() {
    }
    ShareButtonsModule.forRoot = function (config) {
        return {
            ngModule: ShareButtonsModule,
            providers: [
                { provide: core$1.CONFIG, useValue: config },
                {
                    provide: core$1.ShareButtons,
                    useFactory: ShareButtonsFactory,
                    deps: [core$1.CONFIG]
                }
            ]
        };
    };
    return ShareButtonsModule;
}());
ShareButtonsModule.decorators = [
    { type: core.NgModule, args: [{
                declarations: [
                    ShareButtonsComponent,
                    ExpandButtonComponent
                ],
                imports: [
                    button.ShareButtonModule,
                    common.CommonModule
                ],
                exports: [
                    button.ShareButtonModule,
                    ShareButtonsComponent
                ]
            },] },
];
ShareButtonsModule.ctorParameters = function () { return []; };

exports.ShareButtonsFactory = ShareButtonsFactory;
exports.ShareButtonsModule = ShareButtonsModule;
exports.ShareButtonsComponent = ShareButtonsComponent;
exports.ɵa = ExpandButtonComponent;

Object.defineProperty(exports, '__esModule', { value: true });

})));
